Sources/AWSLambdaRuntimeCore/Documentation.docc/Proposals/0001-v2-api.md

```diff
@@ -0,0 +1,905 @@
+cannot override the default implementation. This has proven challenging for users who want to
+[set up global properties before the Lambda starts-up](https://github.com/swift-server/swift-aws-lambda-runtime/issues/265).
+Setting up global properties is required to customize the Swift Logging, Metric and Tracing backend.
+
+#### Non-trivial transition from SimpleLambdaHandler to LambdaHandler
+
+The `SimpleLambdaHandler` protocol provides a quick and easy way to implement a basic Lambda function. It only requires
+an implementation of the `handle` function where the business logic of the Lambda function can be written.
+`SimpleLambdaHandler` is perfectly sufficient for small use-cases as the user does not need to spend much time looking
+into the library.
+
+However, `SimpleLambdaHandler` cannot be used when services such as a database client need to be initialized before the
+Lambda runtime starts and then also gracefully shutdown prior to the runtime terminating. This is because the only way
+to register termination logic is through the `LambdaInitializationContext` (containing a field
+`terminator: LambdaTerminator`) which is created and used _internally_ within `LambdaRuntime` and never exposed through
+`SimpleLambdaHandler`. For such use-cases, other handler protocols like `LambdaHandler` must be used. `LambdaHandler`
+exposes a `context` argument of type `LambdaInitializationContext` through its initializer. Within the initializer,
+required services can be initialized and their graceful shutdown logic can be registered with the
+`context.terminator.register` function.
+
+Yet, `LambdaHandler` is quite cumbersome to use in such use-cases as users have to deviate from the established norms of
+the Swift on Server ecosystem in order to cleanly manage the lifecycle of the services intended to be used. This is
+because the convenient `swift-service-lifecycle` v2 library — which is commonly used for cleanly managing the lifecycles
+of required services and widely supported by many libraries — cannot be used in a structured concurrency manner.
+
+#### Does not integrate well with swift-service-lifecycle in a structured concurrency manner
+
+The Lambda runtime can only be started using the **internal** `Lambda.run()` function. This function is called by the
+`main()` function defined by the `LambdaHandler` protocol, preventing users from injecting initialized services into the
+runtime _prior_ to it starting. As shown below, this forces users to use an **unstructured concurrency** approach and
+manually initialize services, leading to the issue of the user then perhaps forgetting to gracefully shutdown the
+initialized services:
+
+```swift
+struct MyLambda: LambdaHandler {
+    let pgClient: PostgresClient
+
+    init(context: AWSLambdaRuntimeCore.LambdaInitializationContext) async throws {
+        /// Instantiate service
+        let client = PostgresClient(configuration: ...)
```

Comment:
Using Postgres as example about something to use with AWS Lambda is quite misleading for the users for two reasons:
- The main point of using Lambda is autoscaling, this is usually something very difficult to obtain from a RDS like Postgres and could require connection pooling features. Ideally, the use case we should present is the connection to a NO-SQL database such as DynamoDB or MongoDB.
- In addition giving the nature of connection oriented protocol of the DB, opening the client connection during the Lambda's initialisation could lead to connection time-out due to the life cycle of the Lambda. There is no guarantee on how long a lambda will be allocated after the first usage. At some point the connection of the DB will be closed from the DB side, leaving subsequent invocation with a closed connection from the Lambda side.

Reply:
> The main point of using Lambda is autoscaling, this is usually something very difficult to obtain from a RDS like Postgres and could require connection pooling features. Ideally, the use case we should present is the connection to a NO-SQL database such as DynamoDB or MongoDB.

While I generally agree, neither Soto nor MongoDB integrate with swift-service-lifecycle today. The goal of this new API is to integrate better with the shared building blocks of the Swift server ecosystem. Since swift-service-lifecycle is an important one, we opted for PostgresNIO as it integrates with it today.

> In addition giving the nature of connection oriented protocol of the DB, opening the client connection during the Lambda's initialisation could lead to connection time-out due to the life cycle of the Lambda.

PostgresClient can lazily create connections. This is totally configurable.

> At some point the connection of the DB will be closed from the DB side, leaving subsequent invocation with a closed connection from the Lambda side.

That's totally fine, the connection pool will automatically reconnect for the user.